src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -485,14 +471,15 @@ private void buildIndexesBlocking(Collection<SSTableReader> sstables, Set<Index>
 
         try
         {
-            logger.info("Submitting index build of {} for data in {}",
+            logger.info("Submitting index recovery/build of {} for data in {}",
                         indexes.stream().map(i -> i.getIndexMetadata().name).collect(Collectors.joining(",")),
                         sstables.stream().map(SSTableReader::toString).collect(Collectors.joining(",")));
 
             // Group all building tasks
             Map<Index.IndexBuildingSupport, Set<Index>> byType = new HashMap<>();
             for (Index index : indexes)
             {
+                boolean isRecovery = !index.supportsLoad(Loads.ALL);
```

Comment:
This is the only contentious point imo. I don't know if we should call a `Index#isRecoveryNeeded()` method instead rather than trying to infer it at the SIM level. The former assumes `ALL` is mandatory for an index to the 'right', the latter allows for the index implementation to decide that. Both make sense... It's our choice tbh and switching from one to the other in the future it's easy enough anyway.